File: ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoModule.java
/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * <p>This source code is licensed under the MIT license found in the LICENSE file in the root
 * directory of this source tree.
 */
package com.facebook.react.modules.systeminfo;

import static android.content.Context.UI_MODE_SERVICE;

import android.annotation.SuppressLint;
import android.app.UiModeManager;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.provider.Settings.Secure;
import androidx.annotation.Nullable;
import com.facebook.react.R;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.common.build.ReactBuildConfig;
import com.facebook.react.module.annotations.ReactModule;
import com.facebook.react.turbomodule.core.interfaces.TurboModule;
import java.util.HashMap;
import java.util.Map;

/** Module that exposes Android Constants to JS. */
@ReactModule(name = AndroidInfoModule.NAME)
@SuppressLint("HardwareIds")
public class AndroidInfoModule extends ReactContextBaseJavaModule implements TurboModule {
  public static final String NAME = "PlatformConstants";
  private static final String IS_TESTING = "IS_TESTING";

  public AndroidInfoModule(ReactApplicationContext reactContext) {
    super(reactContext);
  }

  /**
   * See:
   * https://developer.android.com/reference/android/app/UiModeManager.html#getCurrentModeType()
   */
  private String uiMode() {
    UiModeManager uiModeManager =
        (UiModeManager) getReactApplicationContext().getSystemService(UI_MODE_SERVICE);
    switch (uiModeManager.getCurrentModeType()) {
      case Configuration.UI_MODE_TYPE_TELEVISION:
        return "tv";
      case Configuration.UI_MODE_TYPE_CAR:
        return "car";
      case Configuration.UI_MODE_TYPE_DESK:
        return "desk";
      case Configuration.UI_MODE_TYPE_WATCH:
        return "watch";
      case Configuration.UI_MODE_TYPE_NORMAL:
        return "normal";
      default:
        return "unknown";
    }
  }

  @Override
  public String getName() {
    return "PlatformConstants";
  }

  @Override
  public @Nullable Map<String, Object> getConstants() {
    HashMap<String, Object> constants = new HashMap<>();
    constants.put("Version", Build.VERSION.SDK_INT);
    constants.put("Release", Build.VERSION.RELEASE);
    constants.put("Serial", Build.SERIAL);
    constants.put("Fingerprint", Build.FINGERPRINT);
    constants.put("Model", Build.MODEL);
    if (ReactBuildConfig.DEBUG) {
      constants.put("ServerHost", AndroidInfoHelpers.getServerHost(getReactApplicationContext().getApplicationContext()));
    }
    constants.put(
        "isTesting", "true".equals(System.getProperty(IS_TESTING)) || isRunningScreenshotTest());
    constants.put("reactNativeVersion", ReactNativeVersion.VERSION);
    constants.put("uiMode", uiMode());
    return constants;
  }

  @ReactMethod(isBlockingSynchronousMethod = true)
  public String getAndroidID() {
    return Secure.getString(getReactApplicationContext().getContentResolver(), Secure.ANDROID_ID);
  }

  @Override
  public void invalidate() {}

  private Boolean isRunningScreenshotTest() {
    try {
      Class.forName("android.support.test.rule.ActivityTestRule");
      return true;
    } catch (ClassNotFoundException ignored) {
      return false;
    }
  }
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/modules/systeminfo/AndroidInfoModule.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.modules.systeminfo

import android.annotation.SuppressLint
import android.app.UiModeManager
import android.content.Context.UI_MODE_SERVICE
import android.content.res.Configuration
import android.os.Build
import android.provider.Settings.Secure
import com.facebook.fbreact.specs.NativePlatformConstantsAndroidSpec
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.common.build.ReactBuildConfig
import com.facebook.react.module.annotations.ReactModule

/** Module that exposes Android Constants to JS. */
@ReactModule(name = NativePlatformConstantsAndroidSpec.NAME)
@SuppressLint("HardwareIds")
@Suppress("DEPRECATION")
public class AndroidInfoModule(reactContext: ReactApplicationContext) :
    NativePlatformConstantsAndroidSpec(reactContext) {

  /**
   * See:
   * https://developer.android.com/reference/android/app/UiModeManager.html#getCurrentModeType()
   */
  private fun uiMode(): String {
    val uiModeManager = reactApplicationContext.getSystemService(UI_MODE_SERVICE) as UiModeManager
    return when (uiModeManager.currentModeType) {
      Configuration.UI_MODE_TYPE_TELEVISION -> "tv"
      Configuration.UI_MODE_TYPE_CAR -> "car"
      Configuration.UI_MODE_TYPE_DESK -> "desk"
      Configuration.UI_MODE_TYPE_WATCH -> "watch"
      Configuration.UI_MODE_TYPE_VR_HEADSET -> "vrheadset"
      Configuration.UI_MODE_TYPE_NORMAL -> "normal"
      else -> "unknown"
    }
  }

  override fun getTypedExportedConstants(): Map<String, Any?> {
    val constants = mutableMapOf<String, Any?>()
    constants["Version"] = Build.VERSION.SDK_INT
    constants["Release"] = Build.VERSION.RELEASE
    constants["Serial"] = Build.SERIAL
    constants["Fingerprint"] = Build.FINGERPRINT
    constants["Model"] = Build.MODEL
    constants["Manufacturer"] = Build.MANUFACTURER
    constants["Brand"] = Build.BRAND
    if (ReactBuildConfig.DEBUG) {
      constants["ServerHost"] =
          AndroidInfoHelpers.getServerHost(reactApplicationContext.applicationContext)
    }
    constants["isTesting"] = "true" == System.getProperty(IS_TESTING) || isRunningScreenshotTest()
    val isDisableAnimations = System.getProperty(IS_DISABLE_ANIMATIONS)
    if (isDisableAnimations != null) {
      constants["isDisableAnimations"] = "true" == isDisableAnimations
    }
    constants["reactNativeVersion"] = ReactNativeVersion.VERSION
    constants["uiMode"] = uiMode()
    return constants
  }

  override fun getAndroidID(): String {
    return Secure.getString(reactApplicationContext.contentResolver, Secure.ANDROID_ID)
  }

  override fun invalidate() {}

  private fun isRunningScreenshotTest(): Boolean {
    return try {
      Class.forName("com.facebook.testing.react.screenshots.ReactAppScreenshotTestActivity")
      true
    } catch (ignored: ClassNotFoundException) {
      false
    }
  }

  public companion object {
    public const val NAME: String = NativePlatformConstantsAndroidSpec.NAME
    private const val IS_TESTING = "IS_TESTING"
    private const val IS_DISABLE_ANIMATIONS = "IS_DISABLE_ANIMATIONS"
  }
}
